### ISSUE TITLE
Cygwin build issues with src/lib/libdll/features/dll and src/cmd/builtin/features/pty.

### ISSUE THREAD

**Issue by ghost** [ksh93/ksh#360]
So, I was trying to get this to build under Cygwin on Windows (I figured a Unix-like environment was my best bet), and I ran into a few problems. I am using Windows Version 10.0.19042.1348 according to ver, if that helps.

The first one I ran into was that there is a script that runs when it's going through src/cmd/builtin/features/pty that just hangs. Like it will sit there forever in some kind of loop until I hit Ctrl-C. I was able to guess that the problem it was having was that it couldn't find the `ptsname` function though, so I looked through the Cygwin headers to see where that was defined, and found it was blocked off under _XOPEN_SOURCE. So I tried adding that to CCFLAGS (along with defining the -d1 debug flag for iffe), and it got past that error.

Log of first build:
[make.out.txt](https://github.com/ksh93/ksh/files/7654389/make.out.txt)

The second problem is much more confusing. Even with debugging enabled, it's not really clear what is going wrong. It just flat out refuses to compile the output block for the dll headers and it's not really clear what is wrong with the code.

Log of second build:
[make.out2.txt](https://github.com/ksh93/ksh/files/7654390/make.out2.txt)

I know that most of the people who would want to use ksh93 are on Unix, so this probably isn't a priority, but I have kind of gotten used to ksh93 on illumos and I'd like to have it on Windows too, since I do still have to use that a lot.

**Comment by hyenias** [comment]
I was able to reproduce the above lockup at pty iffe test as your make.out.txt file indicates using current repo. I am running Win10Pro 21H2 using current cygwin64 with gcc 10.2.

**Comment by McDutchie** [comment]
I can only test on 32-bit Cygwin on a Windows 7 VM. Something we did a long time ago broke the Cygwin build because commit cd5c181a39e14b9785090, which restored pty, builds just fine. I'm now doing a `git bisect` run to find out which commit broke it. This will take a long time as building is glacially slow on this platform.

**Comment by JohnoKing** [comment]
I can reproduce the lockup on Windows 11 64-bit, albeit intermittently. Also, on the occasions ksh does build successfully, the build process is ridiculously slow compared to compiling under any other OS, usually taking about 45 minutes to an hour (even excluding the ksh repo from Windows Defender has no effect).

**Comment by McDutchie** [comment]
Also, by debugging that pty test I found that it locks up on trying to write to stdout. Any `sfprintf(sfstdout,...)` command, even in isolation after deleting everything else, will lock up. This is not good.

**Comment by ghost** [comment]
Right, my guess is that the commit that broke it will be the one that removed the definition of `ptsname` from the code, because it originally had a definition as an `extern char` there in the commit you linked. I was already looking through that history a bit, and I noticed at some point you removed the definition and start including the `<stdlib.h>` header expecting it to be defined there (as it should be), but that's only defined on Cygwin if you build with _XOPEN_SOURCE, and trying that leads me to another failure later on so I'm not 100% sure that is the right fix.

**Comment by McDutchie** [comment]
The thing is, the Cygwin version *is* built with `_XOPEN_SOURCE` (and `_GNU_SOURCE`) as per the generated ast_standards.h file. Except the pty tests don't include that file. To fix that, I added `#include <ast_standards.h>` before the `#include` in src/lib/libast/std/stdio.h, as all the pty tests (implicitly) include that stdio.h. Sure enough, the compiler warning about an implicitly defined ptsname disappeared from the iffe debug output. But that did nothing to fix the lockup on my testing system. So there is more to this.

**Comment by McDutchie** [comment]
Update: The `bisect` run just confirmed that 72968eaed6106 builds fine on Cygwin 32-bit, and that was on Feb 18, well after I removed the explicit `ptsname` extern in 77ab60a149c77f1 on Jan 27 (because it was never actually compiled).

**Comment by McDutchie** [comment]
Forget everything I wrote above, it's all nonsense. I've been chasing ghosts. What a waste of time.

The current master branch builds just fine on Cygwin 32-bit if I pass `CCFLAGS='-O0 -D_std_malloc'`, disabling both compiler optimisation and vmalloc.

So now I only need to figure out if compiler optimisation or vmalloc is the problem.

The regression test results are not pretty at all, by the way. :-/

**Comment by JohnoKing** [comment]
> So now I only need to figure out if compiler optimisation or vmalloc is the problem.

Looking at vmalloc's code it appears `_AST_std_malloc` is defined under Cygwin, with the above comment indicating the intent is to defer to standard malloc (i.e., `_std_malloc` should already have been the default on Windows). The fact this code isn't working properly is probably the problem:
https://github.com/ksh93/ksh/blob/e44436276318230e7fcc5305c7182dd51b0b23ab/src/lib/libast/vmalloc/malloc.c#L82-L90

**Comment by ghost** [comment]
I can confirm that the build flags you used get me a working ksh on Cygwin on 64-bit Windows too. There is one minor hiccup though ksh itself works fine... the build system seems to finish up the build, and then start copying files over to the `arch` directory, all the way up to pushd and popd, and then it just hangs there until I hit Ctrl-C. However, the generated ksh itself seems to work fine. 

Thank you very much for helping me out with this. I never would have guessed it was a malloc issue or compiler optimizations, I would have probably spend a ton of time looking through headers making sure all the includes were there, and defining as many things as I can think of to make sure everything is found.

**Comment by McDutchie** [comment]
Rebuilt ksh on Cygin 32-bit with `CCFLAGS='-Os -D_std_malloc'`, so, optimisation enabled, AST vmalloc disabled. Builds fine. So, optimisation is not the problem, vmalloc is. It's incompatible with cygwin and @JohnoKing's [hunch](https://github.com/ksh93/ksh/issues/360#issuecomment-986170718) is correct: the code is not disabling it properly.

Adding `#define _std_malloc 1` to that `#if` block doesn't help either, it should be disabled in an earlier stage.

Actually, never mind that. Vmalloc has wasted my time for the last time, it's getting disabled by default even on development builds and you'll have to pass `-D_AST_vmalloc` in `CCFLAGS` to enable it for testing purposes.

**Comment by JohnoKing** [comment]
Since vmalloc causes so many bugs (see also https://github.com/ksh93/ksh/issues/95), I have made an experimental branch that removes all traces of vmalloc: https://github.com/JohnoKing/ksh/commit/1d14de67791a9b41156db1dbf21b8348bd551310. Without vmalloc present in any capacity ksh seems to work fine, although there are regress test failures ~(which I suspect indicate memory leaks when using `-D_std_malloc` on the normal repo)~ which occur because of the removal of `vmclose` (which frees regions of memory):
```
test leaks begins at 2021-12-05+08:11:54
	leaks.sh[157]: memory leak on PATH reset before PATH search (leaked approx 4022272 bytes after 4096 iterations)
	leaks.sh[363]: set PATH value in main shell (leaked approx 4022272 bytes after 4096 iterations)
test leaks failed at 2021-12-05+08:11:56 with exit code 2 [ 41 tests 2 errors ]
test leaks(C.UTF-8) begins at 2021-12-05+08:11:56
	leaks.sh[157]: memory leak on PATH reset before PATH search (leaked approx 4018176 bytes after 4096 iterations)
	leaks.sh[256]: associative array in function (leaked approx 77824 bytes after 4096 iterations)
	leaks.sh[363]: set PATH value in main shell (leaked approx 4018176 bytes after 4096 iterations)
test leaks(C.UTF-8) failed at 2021-12-05+08:11:59 with exit code 3 [ 41 tests 3 errors ]
test leaks(shcomp) begins at 2021-12-05+08:11:59
	shcomp-leaks.ksh[197]: memory leak on PATH reset before PATH search (leaked approx 4022272 bytes after 4096 iterations)
	shcomp-leaks.ksh[403]: set PATH value in main shell (leaked approx 4018176 bytes after 4096 iterations)
test leaks(shcomp) failed at 2021-12-05+08:12:02 with exit code 2 [ 41 tests 2 errors ]
```

**Comment by McDutchie** [comment]
Thanks for that, I'll definitely be looking at that, but for now I'll just disable vmalloc by default by defining both `_AST_std_malloc` and `_std_malloc` (unless `_AST_malloc` is passed). In my testing, this does not cause any current leaks.sh tests to fail. So it might be that something still needs tweaking in your branch.

**Comment by hyenias** [comment]
Yesterday was the first time I have ever installed cygwin64 on my win10 box. When compiling ksh93u+m, it also took 45 minutes long which makes it the slowest system I have for compiling ksh. Same hardware running on Ubuntu 18.04 completes a compile for ksh93u+m in less than 5 minutes.

I have done all the things I could find on the Internet to improve performance on my win10 cygwin64 install which has helped bring the compile time from 45 minutes to 30 minutes.

1. Install cygserver and activate it
2. Antimalware Service Executable: Disable temporarily Windows Defender's real-time protection as providing a folder exclusion did not help
3. Temporarily stop SysMain as it was consuming cpu just like antimalware service

I still find the iffe tests exceedingly slow and perhaps something needs to be updated as suggested from [Cygwin's FAQ](https://cygwin.com/faq.html#faq.using.slow).

> Why is Cygwin suddenly so slow?
> If suddenly every command takes a very long time, then something is probably attempting to access a network share. You may have the obsolete //c notation in your PATH or startup files. Using //c means to contact the network server c, which will slow things down tremendously if it does not exist.

When I compiled ksh93u+m with `CCFLAGS='-O0 -D_std_malloc' bin/package make` it also hung up at the pushd statements. However, when I run `bin/shtests` I get 224 errors all over the place (`bin/shtests -p` has 75 errors).

**Comment by McDutchie** [comment]
Cygwin is *incredibly* slow all by itself. There's not much we can do about it.

**Comment by McDutchie** [comment]
@athenian200, after the latest commit it should build correctly by default, without passing any flags. Please let me know if this is the case.

**Comment by ghost** [comment]
It seems to work as described. The build process still mysteriously hangs at the very end (after copying pushd to arch), but the ksh generated works absolutely fine.

**Comment by McDutchie** [comment]
I cannot reproduce that final hanging issue on my cygwin 32-bit build and I don't have a Windows 64-bit installation, so I hope either @JohnoKing or @hyenias can help you debug that…

**Comment by JohnoKing** [comment]
I can't reproduce any sort of hang when building the latest commit on Cygwin, so I don't think I'll be able to help debug that problem.

**Comment by ghost** [comment]
If neither of you can reproduce the issue, I'm just going to assume that last problem is either a bug in my version of Cygwin, or possibly a corrupted git repo. In any case, it's not related to the issue just fixed and I have a working ksh now. So thanks once again for all your help.

**Comment by hyenias** [comment]
So, I turned on my Win10 box after a week to troubleshoot the terminal hang at the last step of `bin/package make` that I was getting. I stopped the SysMain service and turned off the real-time "Virus & threat protection". Then ran the compile and the process completed successfully by returning the console output of "package: make done  at Mon Dec 13 11:24:04 EST 2021 in /home/xx/git/ksh93/ksh/arch/cygwin.i386-64". I tried the exact repo that failed before and it now completes. I tried a fresh ksh93 head and it completes.

I researched into what came after the "+ chmod ugo+x ... pushd" and found it in ./src/cmd/ksh93/Mamfile. It is the end of the loop. I checked out `arch/*/lib/package/gen/make.out` and found that it had completed successfully with writing the timestamp package make completion line even though it hung on doing so to the console window. I verified all the `arch/*/fun/*` files and they were all there with correct permissions.

So, what did I do differently? As my Win10 box was off for a while, I performed a system update. Then, stopped the SysMain services and real-time virus & treat protection before attempting anything CGYWIN related. My guess is that the last two steps were the enabling factor to not have the console window hang during the last output line.

Result: Cleared during testing...
